add new module xcc_redfish_command to manage Lenovo servers using Redfish APIs

panyy3 opened the pull request:
##### SUMMARY
Add this module to manage Lenovo servers using Redfish APIs in order to cover Lenovo oem features that are currently not supported by standard Redfish modules in ansible. 

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
xcc_redfish_command

##### ADDITIONAL INFORMATION
  - Builds Redfish URIs locally and sends them to remote OOB controllers to
    perform an action or get information back or update a configuration attribute.
  - Manages virtual media.
  - Supports getting information back via GET method.
  - Supports updating a configuration attribute via PATCH method.
  - Supports performing an action via POST method.

felixfontein commented on plugins/modules/remote_management/lenovoxcc/xcc_redfish_command.py:
```diff
@@ -0,0 +1,599 @@
+---
+module: xcc_redfish_command
+short_description: Manages Lenovo Out-Of-Band controllers using Redfish APIs
+version_added: 2.2.0
```
2.2.0 has already been released, so nothing can be added to it anymore:
```suggestion
version_added: 2.3.0
```

felixfontein commented on plugins/modules/remote_management/lenovoxcc/xcc_redfish_command.py:
```diff
@@ -0,0 +1,599 @@
+  category:
+    required: true
+    description:
+      - Category to execute on OOB controller
```
```suggestion
      - Category to execute on OOB controller.
```
Please end all descriptions (except `short_description`) with a period (or otherwise use complete sentences).

felixfontein commented on plugins/modules/remote_management/lenovoxcc/xcc_redfish_command.py:
```diff
@@ -0,0 +1,599 @@
+        description:
+          - Indicates if the image is treated as inserted on command completion
+        type: bool
+        default: True
```
```suggestion
        default: true
```
Always use `true`/`false` for booleans in YAML.

felixfontein commented on plugins/modules/remote_management/lenovoxcc/xcc_redfish_command.py:
```diff
@@ -0,0 +1,599 @@
+          - The URL of the image to insert or eject
+        type: str
+      inserted:
+        required: false
```
It is usually a good idea to leave away `required: false`.
```suggestion
```

felixfontein commented on plugins/modules/remote_management/lenovoxcc/xcc_redfish_command.py:
```diff
@@ -0,0 +1,599 @@
+    description: different results depending on task 
+    returned: always
+    type: dict
+    sample: "Action was successful" or list of Lenovo FoD key
```
`msg` is usually a message, and thus a string. So you should fix one type and use that one.

If you want to return something else than a string, use another return value.

felixfontein reviewed (commented):
Thanks for the contribution!

Please note that all new modules must have tests (unit tests and/or integration tests).

BTW, since you seem to work with/for Lenovo: if you plan to create more Lenovo XCC modules, please consider creating an own collection for that. We do not accept new groups of modules from vendors in this collection.

panyy3 commented:
> 
> 
> Thanks for the contribution!
> 
> Please note that all new modules must have tests (unit tests and/or integration tests).
> 
> BTW, since you seem to work with/for Lenovo: if you plan to create more Lenovo XCC modules, please consider creating an own collection for that. We do not accept new groups of modules from vendors in this collection.

felixfontein, Thanks for your review and comment. You are right, I am working for Lenovo, and we just plan to add this one module for lenovo redfish API access. I have updated the code and added unit test script based on your comment. Can you review them again?

felixfontein commented on plugins/modules/remote_management/lenovoxcc/xcc_redfish_command.py:
```diff
@@ -0,0 +1,634 @@
+from ansible_collections.community.general.plugins.module_utils.redfish_utils import RedfishUtils
+
+
+class xcc_RedfishUtils(RedfishUtils):
```
```suggestion
class XCCRedfishUtils(RedfishUtils):
```
In Python, convention is to use camel case for class names.

panyy3 replied:
Good suggestion, I have updated the code.

felixfontein commented on tests/unit/plugins/modules/remote_management/lenovoxcc/test_xcc_redfish_command.py:
```diff
@@ -0,0 +1,653 @@
+import ansible_collections.community.general.plugins.modules.remote_management.lenovoxcc.xcc_redfish_command as module
+
+
+def set_module_args(args):
```
Why do you inline this (and `AnsibleExitJson` and `AnsibleFailJson` etc.) instead of using them from ansible_collections.community.general.tests.unit.plugins.modules.utils?

panyy3 replied:
Thanks for reminder. I didn't realize that these functions have been defined already. I have updated the code.

felixfontein commented:
CC team_redfish (@billdodd @mraineri @tomasg2012 @xmadsen @renxulei) - I know too little about Redfish to know whether this is part of your scope or not, so please take a look and decide :)

billdodd reviewed (commented):
I reviewed the changes. Overall, I think it looks pretty good. A few comments and questions:

1. This OEM extension is structured the way we envisioned for extending the standard Redfish modules. Good job.
2. I noticed that the new xcc_RedfishUtils/XCCRedfishUtils class implements virtual_media_eject() but not virtual_media_insert(). I assume that is because the existing virtual_media_insert() in RedfishUtils works as-is without needing any changes. Is that correct?
3. The standard Redfish modules were very recently updated to allow session login authentication. I expect you would like to be able to take advantage of that in the Lenovo module(s). If so, you should make changes to xcc_redfish_command.py like the ones made to idrac_redfish_command.py seen here: https://github.com/ansible-collections/community.general/pull/2053/files#diff-3394b412b028c7ab6430063e8459c0a3b3ef22a8a9e22c9e779bc931f4724df1
4.  One final topic for consideration/discussion: Who will be the maintainer(s) for the Lenovo module extensions in `plugins/modules/remote_management/lenovoxcc`? @panyy3 Will you and/or other Lenovo employees handle that?

panyy3 commented:
Billdodd, Thanks for your review.
For 2. You are right. The existing virtual_media_insert() in RedfishUtils works as-is without needing any changes. But _find_empty_virt_media_slot() used by virtual_media_insert() need a little change to skip some slot that not support insert. So VirtualMediaInsert need to be added in this module too.
For 3. Thanks for your suggestion. I have updated the code to support using security token for auth.
For 4. I would like to be maintainer of this module, if possible. If existing maintainer is preferred, my colleague renxulei can be maintainer.

renxulei reviewed (commented):
@panyy3 Thanks for your contribution. It looks good for me. I saw Raw commands in this module are common implementation, no specific codes combined with Lenovo, suggest you can add it in standard Redfish module which can benefit more users in future.

billdodd commented:
> Billdodd, Thanks for your review.
> For 2. You are right. The existing virtual_media_insert() in RedfishUtils works as-is without needing any changes. But _find_empty_virt_media_slot() used by virtual_media_insert() need a little change to skip some slot that not support insert. So VirtualMediaInsert need to be added in this module too.
> For 3. Thanks for your suggestion. I have updated the code to support using security token for auth.
> For 4. I would like to be maintainer of this module, if possible. If existing maintainer is preferred, my colleague renxulei can be maintainer.

Sounds good. Thanks!

felixfontein commented on plugins/modules/remote_management/lenovoxcc/xcc_redfish_command.py:
```diff
@@ -0,0 +1,673 @@
+---
+module: xcc_redfish_command
+short_description: Manages Lenovo Out-Of-Band controllers using Redfish APIs
+version_added: 2.3.0
```
```suggestion
version_added: 2.4.0
```

panyy3 replied:
Version has been updated to 2.4.0

felixfontein reviewed (commented):
2.3.0 got released yesterday, so this needs a bump. (2.4.0 will be released next week, so it won't be long :) )

felixfontein commented:
> For 4. I would like to be maintainer of this module, if possible. If existing maintainer is preferred, my colleague renxulei can be maintainer.

There can also be multiple maintainers :) We can also add team_redfish as maintainers (next to you), if team_redfish is interested in this (this would implicitly include @renxulei). Or if the others don't want to, we can add both you and @renxulei (assuming he agrees).

To add yourself (and potentially others), add two lines around here: https://github.com/ansible-collections/community.general/blob/main/.github/BOTMETA.yml#L704 
```
  $modules/remote_management/lenovoxcc/:
    maintainers: panyy3
```
and potentially more names (space-separated) to `maintainers`.

felixfontein commented on .github/BOTMETA.yml:
```diff
@@ -703,6 +703,8 @@ files:
     maintainers: bgaifullin cloudnull
   $modules/remote_management/lxca/:
     maintainers: navalkp prabhosa
+  $modules/remote_management/lenovoxcc/:
```
Can you move this before the previous entry, so it's sorted alphabetically? Thanks :)

panyy3 replied:
Sorry, I haven't notice it. Code is updated.

felixfontein commented:
@renxulei @billdodd I think all formal aspects are ok, so if you're happy with this being merged, please add an approving review. I'll then merge.

panyy3 commented:
> 
> 
> @panyy3 Thanks for your contribution. It looks good for me. I saw Raw commands in this module are common implementation, no specific codes combined with Lenovo, suggest you can add it in standard Redfish module which can benefit more users in future.

Good suggestion. I will try to add it in future.

renxulei reviewed (approved):
Shipit

billdodd reviewed (approved):
@panyy3 Thanks for the contribution.

LGTM

shipit

felixfontein commented:
bot_status

felixfontein commented:
@panyy3 thanks for contributing this new module!
@billdodd @renxulei thanks for reviewing this!
